Add package which sorts input string with integers splited by spaces

mswatermelon opened the pull request:
https://gb.ru/lessons/182480/homework

mswatermelon committed:
Add package whick sorts input string with integers splited by spaces

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,55 @@
+
+		for i := 1; i < len(numbers); i++ {
+			number := numbers[i]
+			numberToCompare := i
```
здесь немного есть несоответствеие в названиях переменных. 
Две переменные называются `number` и `numberToCompare`. Кажется что это 2 каких-то числа, каких-то величины одного и того же порядка (то есть их можно быть сравнивать)
На самом деле  - это абсолютно разные сущности: number это число (элемент массив), а numberToCompare  - это индекс.

Подумайте, как лучше называть переменные здесь

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,55 @@
+		if len(numbers) != 0 {
+			fmt.Println("Result:")
+		}
+		for i := 0; i < len(numbers); i++ {
```
```
fmt.Println(strings.Join(numbers, " "))
```

вместо всего цикла :)

mswatermelon replied:
Мне нужно где-то преобразовать каждое число в строку прежде чем воспольховаться join. На ум приходит либо опять цикл, либо делать это в функции сортировки
```
func sort(numbers []string) {
	for i := 1; i < len(numbers); i++ {
		number := numbers[i]
		var firstNumber, secondNumber int64
		var err error
		if firstNumber, err = strconv.ParseInt(number, 10, 64); err != nil {
			fmt.Println("Error happend during parcing, try again")
			continue
		}
		j := i
		if secondNumber, err = strconv.ParseInt(numbers[j-1], 10, 64); err != nil {
			fmt.Println("Error happend during parcing, try again")
			continue
		}
		for ; j >= 1 && secondNumber > firstNumber; j-- {
			numbers[j] = numbers[j-1]
		}
		numbers[j] = number
	}
}
```

goarchitecture replied:
Упс, да, не то. Тут у вас слайс интеджеров, а не слайс строк.
В принципе ваш цикл - вполне возможен, но получается сликшом громоздкая конструкция всего лишь для вывода информации. Тут есть тогда варианты:
1) вывести в отдельную функцию pintResult()
2) вывести просто через fmt.Println("%v", numbers) (но тогда будут `[]` - квадратные скобки по бокам)
3) конвертировать в массив строк (можно отдельной строкой) и вывести через join

goarchitecture replied:
P.S. - нет, сортировать []string - не имеет смысл. Вы же всё-таки сортируете по численному принципу.
Поэтому сортировать надо `[]int`-подобный тип.
Даже, если вдруг вам нужно всё-таки переводить это в массив строк для вывода, не смешивайте логику и представление. Логика - в данном случае сортировка. Она была выполнена и работала.
А представление - это отдельно. Отдельная функция, которая "выводит массив" на экран, и по сути этой функции всё равно где этот массив взялся (после сортировки, или любой другой функции)

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,55 @@
+	scanner := bufio.NewScanner(os.Stdin)
+
+	fmt.Println("Enter integers to sort splitted by space and press Enter")
+	for scanner.Scan() {
```
что ваш цикл на самом деле здесь делает?
пользователю выводится сообщение "введите числа разделенные пробелом и нажмите энтер" - программа выведет отсортированное, и вы начинаете заново.. но как пользователь узнает что можно снова вводить числа?
и как пользователю выйти?

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,55 @@
+				var err error
+				if parsedNumber, err = strconv.ParseInt(number, 10, 64); err != nil {
+					fmt.Println("Error happend during parcing")
+					os.Exit(1)
```
если уже есть верхний цикл, то может "мягко" пропускать и начинать по циклу снова? (но с сообщение пользователю конечно)

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,55 @@
+			}
+		}
+
+		for i := 1; i < len(numbers); i++ {
```
вынесите саму сортировку в отдельную функцию

goarchitecture reviewed (changes requested):
Хороший старт. Доработайте по комментариям

mswatermelon committed:
Use separate function to print result and sort, implement correct exit from loop, fix variable's names, continue with a new loop, if there is an error on parse

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,59 @@
+				var parsedNumber int64
+				var err error
+				if parsedNumber, err = strconv.ParseInt(number, 10, 64); err != nil {
+					fmt.Println("Error happend during parcing, try again")
```
напишите номер числа в котором произошла ошибка, или само число. А то пользователь может и не понимать в чём именно дело

goarchitecture commented on lesson4/sort.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+func convertSliceToString(sliceToConvert []int64, delim string) string {
+	return strings.Trim(strings.Replace(fmt.Sprint(sliceToConvert), " ", delim, -1), "[]")
```
:+1:

goarchitecture reviewed (commented):
отлично!
